src/context.rs

```diff
@@ -468,6 +401,65 @@ impl<A: Actor> Context<A> {
 
         Ok(fut)
     }
+
+    fn inbox(&self) -> Inbox<A> {
+        Inbox::new(self.receiver.clone(), self.broadcast_receiver.clone())
+    }
+}
+
+struct Inbox<A> {
+    inner: BoxStream<'static, Either<BroadcastMessage<A>, AddressMessage<A>>>,
+}
+
+impl<A> Inbox<A>
+where
+    A: 'static,
+{
+    fn new(
+        address_receiver: Receiver<AddressMessage<A>>,
+        broadcast_receiver: barrage::SharedReceiver<BroadcastMessage<A>>,
+    ) -> Self {
+        let broadcast_stream =
+            futures_util::stream::unfold(broadcast_receiver, |broadcast_receiver| async move {
+                match broadcast_receiver.recv_async().await {
+                    Ok(msg) => Some((msg, broadcast_receiver)),
+                    Err(_) => None,
+                }
+            })
+            .map(Either::Left)
+            .boxed();
+
+        let address_stream =
+            futures_util::stream::unfold(address_receiver, |address_receiver| async move {
+                match address_receiver.recv_async().await {
+                    Ok(msg) => Some((msg, address_receiver)),
+                    Err(_) => None,
+                }
+            })
+            .map(Either::Right)
+            .boxed();
+
+        Self {
+            inner: futures_util::stream_select!(broadcast_stream, address_stream).boxed(),
+        }
+    }
+
+    pub async fn next(&mut self) -> Option<InboxMessage<A>> {
+        let inner = self.inner.next().await?;
+
+        Some(InboxMessage { inner })
+    }
+}
+
+struct InboxMessage<A> {
+    inner: Either<BroadcastMessage<A>, AddressMessage<A>>,
```

Comment:
I wonder if it'd be simpler to just merge BroadcastMessage<A> and AddressMessage<A> as such:

```rust
enum BroadcastMessage<A: Actor> {
    Message(Box<dyn MessageEnvelope<Actor = A>>),
    Broadcast(Box<dyn BroadcastMessageEnvelope<Actor = A>>),
    Shutdown,
    LastAddress,
}
```

Reply:
Yeah that sounds like a good idea to me! I'd still wrap it in `InboxMessage` though so that enum details remain an implementation detail of the library.

Reply:
> I wonder if it'd be simpler to just merge BroadcastMessage and AddressMessage as such:
> 
> ```rust
> enum Message<A: Actor> {
>     Message(Box<dyn MessageEnvelope<Actor = A>>),
>     Broadcast(Box<dyn BroadcastMessageEnvelope<Actor = A>>),
>     Shutdown,
>     LastAddress,
> }
> ```

Unfortunately it is not that easy because `barrage::Sender` for example enforces that `T` is `Clone` and `MessageEnvelope` cannot be cloned. We could combine it _just_ for whatever is in `InboxMessage` but I don' think it is worth it.

Reply:
Ah, right. Unfortunate, we can't really relax the Clone requirement either as it is not possible conceptually to have broadcast without Cloning (or an ugly solution involving `(FnOnce() -> T)` + Clone closures)